File: resources/code/view_model/view_model.py
from libcellml import Model, Component, Variable, Reset


def main():
    # Create a model and populate it with components, variables, resets, etc.
    my_model = Model("Grandfather")

    my_uncle_component = Component("Uncle")
    my_model.addComponent(my_uncle_component)

    my_component = Component("Mother")
    my_model.addComponent(my_component)

    my_daughter_component = Component("Daughter")
    my_component.addComponent(my_daughter_component)

    my_son_component = Component("Son")
    my_component.addComponent(my_son_component)

    my_grandchild_component = Component("Grandchild")
    my_son_component.addComponent(my_grandchild_component)

    # Creating a Variable instance:
    my_variable = Variable("testVariable")
    my_variable.setInitialValue("3.14")
    my_component.addVariable(my_variable)

    my_reset = Reset()
    my_reset.setTestVariable(my_component.variable("testVariable"))
    my_component.addReset(my_reset)

    # start - VIEW_MODEL 1

    # Retrieve the model name.
    my_model_name = my_model.name()

    # Retrieve the initial value of the variable.
    my_initial_value = my_variable.initialValue()

    # end - VIEW_MODEL 1

    # start - VIEW_MODEL 2

    # Retrieve the test variable from the reset.
    my_test_variable = my_reset.testVariable()

    # end - VIEW_MODEL 2

    # start - VIEW_MODEL 3

    # Loop through variables in the component and retrieve names
    for v in range(my_component.variableCount()):
        my_variable = my_component.variable(v)
        my_variable_name = my_variable.name()

    # Retrieve a variable by name.
    my_hello_there_variable = my_component.variable("helloThere")

    # This will return None if not found.
    my_missing_variable = my_component.variable("nameThatDoesntExist")

    # This will cause an error.
    # my_missing_name = my_missing_variable.name()

    # Safe alternative:
    if my_missing_variable is not None:
        my_missing_name = my_missing_variable.name()

    # end - VIEW_MODEL 3

    grandfather = my_model

    # start - VIEW_MODEL 4

    # Direct children count only.
    grandfather_has_two_kids = grandfather.componentCount()  # 2

    # Check children of components.
    uncle_has_no_kids = grandfather.component(0).componentCount()  # 0
    mother_has_two_kids = grandfather.component("Mother").componentCount()  # 2

    # end - VIEW_MODEL 4

    return 0


if __name__ == "__main__":
    exit(main())

File: resources/code/view_model/view_model.cpp
#include <libcellml>

int main(int argc, char *argv[])
{
    // Create a model and populate it with components, variables, resets, etc. 
    libcellml::ModelPtr myModel = libcellml::Model::create("Grandfather");
    libcellml::ComponentPtr myUncleComponent = libcellml::Component::create("Uncle");
    myModel->addComponent(myUncleComponent);
    libcellml::ComponentPtr myComponent = libcellml::Component::create("Mother");
    myModel->addComponent(myComponent);
    libcellml::ComponentPtr myDaughterComponent = libcellml::Component::create("Daughter");
    myComponent->addComponent(myDaughterComponent);
    libcellml::ComponentPtr mySonComponent = libcellml::Component::create("Son");
    myComponent->addComponent(mySonComponent);
    libcellml::ComponentPtr myGrandchildComponent = libcellml::Component::create("Grandchild");
    mySonComponent->addComponent(myGrandchildComponent);

    // Creating a Variable instance:
    libcellml::VariablePtr myVariable = libcellml::Variable::create("testVariable");
    myVariable->setInitialValue("3.14");
    myComponent->addVariable(myVariable);

    libcellml::ResetPtr myReset = libcellml::Reset::create();
    myReset->setTestVariable(myComponent->variable("testVariable"));
    myComponent->addReset(myReset);

    // start - VIEW_MODEL 1

    // Retrieving the name attribute from the myModel item:
    const std::string myModelName = myModel->name();

    // Retrieving the initial value of the myVariable item:
    const std::string myInitialValue = myVariable->initialValue();

    // end - VIEW_MODEL 1

    // start - VIEW_MODEL 2

    // Retrieve the test variable from the myReset item:
    libcellml::VariablePtr myTestVariable = myReset->testVariable();

    // end - VIEW_MODEL 2

    // start - VIEW_MODEL 3

    // Loop through variables in the component myComponent and retrieve their names.
    //    NOTE that indexing starts from zero.
    for(size_t v = 0; v < myComponent->variableCount(); ++v) {

        // Retrieve the Variable item at index v:
        auto myVariable = myComponent->variable(v);

        // Retrieve the name of the myVariable item:
        auto myVariableName = myVariable->name();
    }

    // Retrieve a variable called "helloThere" by name.
    //    NOTE that a nullptr will be returned if no variable of that name is found.
    auto myHelloThereVariable = myComponent->variable("helloThere");

    // In this case, the myMissingVariable will be a nullptr:
    auto myMissingVariable = myComponent->variable("nameThatDoesntExist");

    // This will cause a segfault as myMissingVariable is null:
    //auto myMissingName = myMissingVariable->name();

    // Safe alternative:
    if (myMissingVariable != nullptr) {
        auto myMissingName = myMissingVariable->name();
    }

    // end - VIEW_MODEL 3

    auto grandfather = myModel;

    // start - VIEW_MODEL 4

    // The number of components owned by the grandfather model refers *only* to its direct children:
    auto grandfatherHasTwoKids = grandfather->componentCount(); // returns 2

    // Each component must be interrogated individually to determine its children.
    //    Note that the uncle component is the 0th child of the grandfather model.
    auto uncleHasNoKids = grandfather->component(0)->componentCount();          // returns 0
    auto motherHasTwoKids = grandfather->component("Mother")->componentCount(); // returns 2

    // end - VIEW_MODEL 4
    return 0;
  }
